src/hotspot/share/opto/predicates.hpp

```diff
@@ -910,4 +918,70 @@ class Predicates : public StackObj {
 #endif // NOT PRODUCT
 };
 
+// This class checks whether a node is in the original loop body and not the cloned one.
+class NodeInOriginalLoopBody : public NodeInLoopBody {
+  const uint _first_node_index_in_cloned_loop_body;
+  const Node_List& _old_new;
+
+ public:
+  NodeInOriginalLoopBody(const uint first_node_index_in_cloned_loop_body, const Node_List& old_new)
+      : _first_node_index_in_cloned_loop_body(first_node_index_in_cloned_loop_body),
+        _old_new(old_new) {}
+  NONCOPYABLE(NodeInOriginalLoopBody);
+
+  // Check if 'node' is not a cloned node (i.e. "< _first_node_index_in_cloned_loop_body") and if we've created a
+  // clone from 'node' (i.e. _old_new entry is non-null). Then we know that 'node' belongs to the original loop body.
+  bool check(Node* node) const override {
+    if (node->_idx < _first_node_index_in_cloned_loop_body) {
+      Node* cloned_node = _old_new[node->_idx];
+      return cloned_node != nullptr && cloned_node->_idx >= _first_node_index_in_cloned_loop_body;
+    } else {
+      return false;
+    }
+  }
+};
+
+// Visitor to create Initialized Assertion Predicates at a target loop from Template Assertion Predicates from a source
+// loop. This visitor can be used in combination with a PredicateIterator.
+class AssertionPredicatesForLoop : public PredicateVisitor {
+  Node* const _init;
+  Node* const _stride;
+  Node* const _old_target_loop_entry;
+  Node* _new_control;
+  PhaseIdealLoop* const _phase;
+  ParsePredicateSuccessProj* _current_parse_predicate;
```

Comment:
It looks to me like this could be a boolean, correct?

Reply:
Then the name could also be more descriptive

Reply:
I changed it into a `bool`. The idea is to simulate the old behavior that we only create the Assertion Predicates if there are Parse Predicates available. I don't think this is generally correct - we could still split loops later when Parse Predicates have already been removed. But I want to fix this at other places as well and thus defer this to a separate change later.